internal/step/plugin/provider.go

```diff
@@ -885,47 +1026,66 @@ func (r *runningStep) deployStage() (deployer.Plugin, error) {
 	return container, nil
 }
 
-func (r *runningStep) startStage(container deployer.Plugin) error {
-	r.logger.Debugf("Starting stage for step %s/%s", r.runID, r.pluginStepID)
-	atpClient := atp.NewClientWithLogger(container, r.logger)
+// enableStage returns the result of whether the stage was enabled or not.
+func (r *runningStep) enableStage() (bool, error) {
 	r.lock.Lock()
 	previousStage := string(r.currentStage)
-	r.currentStage = StageIDStarting
-	inputRecievedEarly := false
-	r.atpClient = atpClient
-
-	var runInput any
-	select {
-	case runInput = <-r.runInput:
-		// Good. It received it immediately.
-		r.state = step.RunningStepStateRunning
-		inputRecievedEarly = true
-	default: // The default makes it not wait.
-		r.state = step.RunningStepStateWaitingForInput
-	}
-
-	runInputAvailable := r.runInputAvailable
+	r.currentStage = StageIDEnabling
+	enabledInputAvailable := r.enabledInputAvailable
+	r.state = step.RunningStepStateWaitingForInput
 	r.lock.Unlock()
 
 	r.stageChangeHandler.OnStageChange(
 		r,
 		&previousStage,
 		nil,
 		nil,
-		string(StageIDStarting),
-		runInputAvailable,
+		string(StageIDEnabling),
+		enabledInputAvailable,
 		&r.wg,
 	)
 
+	select {
+	case enabled := <-r.enabledInput:
+		return enabled, nil
+	case <-r.ctx.Done():
+		return false, fmt.Errorf("step closed while determining enablement status")
+	}
+}
+
+func (r *runningStep) startStage(container deployer.Plugin) error {
+	r.logger.Debugf("Starting stage for step %s/%s", r.runID, r.pluginStepID)
+	atpClient := atp.NewClientWithLogger(container, r.logger)
+	var inputReceivedEarly bool
 	r.lock.Lock()
-	r.currentStage = StageIDStarting
-	r.logger.Debugf("Waiting for input state while starting 2.")
+	r.atpClient = atpClient
 	r.lock.Unlock()
 
+	var runInput any
+	var newState step.RunningStepState
```

Comment:
Iff you make a change for some other reason, it would be nicer to have the `var` declaration of `inputReceivedEarly` placed with the one for `newState`, since they are managed together in the subsequent code.  (That is, there is no need for the `inputReceivedEarly` declaration until we reach the `switch` statement at line 1066, so having it precede the `Lock()` sequence is just noisy/distracting.)